### PR TITLE
PdfOutline (Outline/Bookmarking functionality)

### LINKED ISSUE #450: [Feature Request] API to create an outline (table of contents) for PDF readers
This should allow users to display a visual outline that is linked to different parts of the pdf.

### PR DESCRIPTION
**Pull request by aready87**
Hi @Hopding,

We put together PDFOutline that adds the outlines (aka bookmarks, table of contents) functionality as specified in Adobe's PDF Specification Section 12.3.3 `/Outlines`. We adhered to the PDF Specification and also how PDFLib was organized and structured as closely as possible (It is structured very closely to PDFPage that wraps PDFPageLeaf; PDFOutline wraps PDFOutlines). 

Scratchpad's index.ts pretty much captures the functionalities that we've added. Notably:
- PDFDocument's CreateOptions now has useOutlines boolean which adds the PageMode UseOutlines for displaying the outline panel when the document is opened (Table 28, pg 74)
- PDFDocument's addOutline and insertOutline adds a top-level outline (PDFOutline) to the document, with option: expanded (for expanded display), linkPage (link the outline to a given PDFPage) or linkIndex (link the outline to a given index). This returns a PDFOutline.
- PDFOutline (once received from the above point) also has a similar addOutline and insertOutline that adds a nested outline to said PDFOutline with the same options (expanded, linkPage and linkIndex.
- PDFOutline also has the following: 
 i) setTitle to change the label/title of the outline
 ii) linkPage to add/change the outline's link to a given PDFPage
 iii) linkIndex to add/change the outline's link to a given index
 iv) setExpanded to change the expanded boolean
 v) remove to delete the outline and its children

We also tried to provide as much comments where necessary, following the style and flow of the other comments in the code (including a shoutout to another library).

This will resolve #450 and is a more complete solution to #127, and hopefully makes it easier to implement a feature request #218.

This passes scratchpad:start without issue and passes `yarn typecheck`, `yarn test` and successfully builds with `yarn build`. But it fails the integration tests. Could you please help us in implementing PDFOutline and resolving the errors? There's likely more to discuss and more to complete (e.g. writing tests and documentation) but we'll give you an opportunity to review the code first and approve the overall structure. Feel free to reach out to us.

Thank you.

-Real @closin-inc

### CODE REVIEW

**Comment by Hopding** [comment]
Hello @aready87! Thanks for working on this! It'll be a great feature to have in the library. I'll try to review this over the weekend and also let you know why the tests are failing.

**Comment by Hopding** [comment]
@aready87

I ran the second integration test `yarn apps:node 2` and encountered this error:
```
...
(node:5739) UnhandledPromiseRejectionWarning: TypeError: rootOutline.endOutline is not a function
    at PDFDocument.endOutline (/Users/user/github/pdf-lib/cjs/api/PDFDocument.js:1198:25)
    at PDFDocument.<anonymous> (/Users/user/github/pdf-lib/cjs/api/PDFDocument.js:1259:51)
    at step (/Users/user/github/pdf-lib/node_modules/tslib/tslib.js:139:27)
    at Object.next (/Users/user/github/pdf-lib/node_modules/tslib/tslib.js:120:57)
    at /Users/user/github/pdf-lib/node_modules/tslib/tslib.js:113:75
    at new Promise (<anonymous>)
    at Object.__awaiter (/Users/user/github/pdf-lib/node_modules/tslib/tslib.js:109:16)
    at PDFDocument.save (/Users/user/github/pdf-lib/cjs/api/PDFDocument.js:1250:24)
    at /Users/user/github/pdf-lib/apps/node-build/tests/test2.js:154:45
    at step (/Users/user/github/pdf-lib/apps/node-build/tests/test2.js:44:23)
...
```

What's noteworthy about the second integration test is that it modifies a document with `PDFDocument.load(...)` instead of creating one with `PDFDocument.create()`. Moreover, the document it modifies already has an outline.

The error occurs in `PDFDocument.endOutline`:
```js
endOutline(): void {
  const rootOutline = this.catalog.Outlines();
  if (rootOutline) {
    rootOutline.endOutline();
  }
}
```

Here, `rootOutline` is defined, but it turns out to be a `PDFDict` rather than a `PDFOutline`. So, of course, the call to `endOutline()` fails because no such method exists on `PDFDict`.

If we look in `PDFCatalog.Outlines()`, we can see why we're not getting the type we expect:
```js
Outlines(): PDFOutlines {
  return this.lookup(PDFName.Outlines, PDFDict) as PDFOutlines;
}
```

At runtime, the only guarentee we have is that the lookup returns an instance of `PDFDict` (or else throws an error). Despite this, we cast it (only applicable at compile time) to a `PDFOutlines`. But this type cast does not hold in all cases (as evidenced by the error above).

The `PDFCatalog.Pages()` method does a similar sort of type cast:

```js
Pages(): PDFPageTree {
  return this.lookup(PDFName.of('Pages'), PDFDict) as PDFPageTree;
}
```

This cast does hold reliably. The reason is because the parser makes sure to create an instance of `PDFPageTree` instead of `PDFDict` whenever it parses a dictionary with `/Type /Pages`:
https://github.com/Hopding/pdf-lib/blob/9b6fb96deaaeb4381c573e0e5f8e9d56629ced4a/src/core/parser/PDFObjectParser.ts#L193-L203

This same sort of pattern (checking the `/Type` entry in the parser and creating specialized subtypes of `PDFDict`) is used for a few other types as well. Unfortunately, this pattern does not scale well at all. It depends on the presence of a `/Type` entry in all dictionaries of a particular type. The specification requires that certain dict types have this entry (such as catalogs, page trees, and page leafs). However, most types of dictionaries either do not have this entry at all or it is optional. This is the case for `Outlines` dictionaries (notice the _(Optional)_ parenthetical):

<kbd>

</kbd>

>

This problem necessitates a different approach. I'm working on an AcroForms API in the [AcroForms](https://github.com/Hopding/pdf-lib/tree/AcroForms) branch where I ran into this issue very quickly. The approach I took to solve it was to use composition instead of inheritance. You can see this approach in my implementation of [PDFAcroForm](https://github.com/Hopding/pdf-lib/blob/AcroForms/src/core/acroform/PDFAcroForm.ts) and the associated getter methods on `PDFCatalog`:
https://github.com/Hopding/pdf-lib/blob/5eee5dbe679de95f0c368c222a0b61ac1d3896a5/src/core/structures/PDFCatalog.ts#L26-L34

There are many other dictionary types I followed this pattern for in the `AcroForms` branch. I've found it to work quite well. So I think that's the best pattern to use for `Outlines` as well. I hope this helps. Please let me know if you have any questions or trouble refactoring your work here to use this compositional pattern!

> Side note: Ultimately I'd like to eliminate the `PDFDict` subclassing pattern altogether from the `pdf-lib` codebase, but doing so will require making breaking changes. So it'll have to wait until `v2.0.0`.

**Comment by Hopding** [comment]
@aready87 I think your approach makes sense overall. It looks like a nice ergonomic API!

However, there are two things I would like to see changed:
1. There are a couple of spots where the user can pass either `linkIndex: number` or `linkPage: PDFPage`. I think I'd prefer to remove the `linkIndex` option and just require the user to always pass a `PDFPage`, at which point the option could be called `page`. I think this API is still quite ergonomic. For example:

    ```js
    outline.addOutline('FooBar', { expanded: true, page: pdfDoc.getPage(2) });
    ```
    The reason I prefer this is that it leaves only one way for the user to accomplish the task. This is a pattern I've tried to follow in designing pdf-lib's API. It tends to be simpler to document and maintain. It can also avoid awkward precendence rules (such as does `linkPage` or `linkIndex` apply if both are supplied?).
2. Instead of passing `useOutlines` to `PDFDocument.create()` I'd prefer to have a `PDFDocument.setDisplayMode()` method. This could work as follows:

    ```js
    enum DisplayMode { ... }

    pdfDoc.setDisplayMode(DisplayMode.None);
    pdfDoc.setDisplayMode(DisplayMode.ShowOutlines);
    pdfDoc.setDisplayMode(DisplayMode.ShowThumbnails);
    pdfDoc.setDisplayMode(DisplayMode.FullScreen);
    pdfDoc.setDisplayMode(DisplayMode.ShowOptionalContent);
    pdfDoc.setDisplayMode(DisplayMode.ShowAttachments);

    // Could also have a `getDisplayMode()` method - although that isn't
    // necessary for this PR
    const displayMode: DisplayMode = pdfDoc.getDisplayMode();
    ```
    This approach allows users to set any type of display mode they want, and it also works for existing documents (with `PDFDocument.load(...)`) not just new ones.

**Comment by aready87** [comment]
@Hopding : Thanks for the detailed explanation! We agree your suggested use of the linkPage API is ergonomic. We will go ahead and remove the linkIndex.

As for the rest, we will review your comments more closely and come back with questions/resolutions.

**Comment by aready87** [comment]
@Hopding: Hi Hopding, hope you've been well. Apologies for the delayed response; We've been quite busy with beta testing our core product. I went ahead and applied the two requested changes: 1) to the options `{ page: PDFPage}` and 2) including `PDFDocument.setDisplayMode(DisplayMode)`.

I am however having difficulty with the suggested compositional pattern without breaking the root outlines' functionality. I just spun my wheel for the last few hours without much success. Sadly I won't be able to get back to this until August (possibly even later); so if you or anyone else want to take a stab at the compositional pattern, I would actually be grateful. Apologies for leaving it unfinished for the time being.

**Comment by Hopding** [comment]
@aready87 No worries! Thanks for working on this! (especially since you've been busy with testing your product)

I'm deep in the weeds implementing a form creation/filling API for pdf-lib, so I'm afraid I don't have time to pick this up right now. If you're able to explain what difficulties you ran into I'd be happy to help you (or anybody else) figure out a solution though.

File: src/core/structures/PDFOutlines.ts [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    context: PDFContext,
+    options?: outlineOptions,
+    title?: string,
+    parent?: PDFRef,
+  ) => {
+    const dict = new Map();
+    if (!parent) {
+      dict.set(PDFName.Type, PDFName.Outlines);
+    } else {
+      dict.set(PDFName.Title, title ? PDFString.of(title) : title);
```
Comment (suninuni) [inline src/core/structures/PDFOutlines.ts:28]: https://github.com/Hopding/pdf-lib/issues/516#issuecomment-656958214

Comment (divinehawk) [reply src/core/structures/PDFOutlines.ts:28]: Setting the outline title with to a string with an unmatched parentheses causes a corrupt PDF.
Per #516 , using PDFHexString appears to fix this problem.

https://github.com/divinehawk/pdf-lib/commit/57ace824d3492bae1705818713a7a1dbc8a80c06

**Comment by drom** [comment]
A very important feature to have! Any plans merging it?

**Comment by Hopding** [comment]
@drom Agreed! I'd very much like to get this included in the library. All that remains to do so is the completion of the compositional pattern explained above (https://github.com/Hopding/pdf-lib/pull/486#issuecomment-650517441). I do not personally have time to work on this yet. If you're interested in helping out, that would definitely speed things along.

**Comment by drom** [comment]
@Hopding I would love to help, but after multiple attempts, my brain was not able to comprehend the sheer design complexity of this library. So as a pacient user I would just wait for a resolution.

**Comment by jainshubham97** [comment]
Hi @Hopding @aready87 , Thanks for having this wonderful feature in the pipeline. However, I need to add hyperlinks to pages in the PDF and I believe this PR has a sub-component (linkPages) that can help me out. Is there any way I can extract it and add it to my library manually? This feature would take time and I am not capable enough to work on the compositional patterns.  Appreciate the help.

**Comment by Hopding** [comment]
@jainshubham97 I’d suggest taking a look at https://github.com/Hopding/pdf-lib/issues/123#issuecomment-568804606. It has a working example showing how to create links to pages in a document

**Comment by jainshubham97** [comment]
@Hopding Thanks brother! Worked like a charm. It saved me many days of effort. I will try to add a high-level API for this if I could as mentioned in #555 (Feature)

**Comment by Hopding** [comment]
@jainshubham97 Glad it worked for you! I look forward to your PR. Let me know if you have any questions.

**Comment by Hopding** [comment]
@Nolence it is indeed very close and I'd love to see it move across the finish line. I was hoping to do the work myself, but I've just got to much on my plate at the moment. Happy to answer questions if anybody is interested in finishing it up though!

If anybody does decide to pick this up, please note that you'll need to either update this PR (via a force push) or open a new PR with your changes. The tl;dr is that I recently had to do a fairly extensive git history rewrite that caused a bunch of conflicts. Please read https://github.com/Hopding/pdf-lib#git-history-rewrite for details. Apologies for the hassle!

**Comment by Hopding** [comment]
Closing this for now. Added outline support to the roadmap in https://github.com/Hopding/pdf-lib/discussions/998.

**Pull request closed without merging**